File: ventures/0016-hedgefonds-simulation-echte-weltwirtschaft/messung-stack/python/schritt.py
#!/usr/bin/env python3
"""Deterministischer Festkomma-Weltschritt -- Umsetzung in Python.

Vorgabe: ../AUFGABE.md. Kein Gleitkommatyp, keine Fremdbibliothek, keine
Argumente, keine Eingabe.

Pythons int ist beliebig genau. Das erfuellt die 128-Bit-Forderung an das
Zwischenprodukt von selbst, verlangt aber umgekehrt, den 64-Bit-Wertebereich
ausdruecklich herzustellen -- dafuer gibt es `umbruch64`.
"""

import time

GROESSE = 64
SCHRITTE = 1_000_000
NACHBAR_VERSATZ = 17
STREUER = 977
NENNER = 10_000
NACHBAR_TEILER = 1_024
EIGEN_TEILER = 4_096
UNTERGRENZE = 0
OBERGRENZE = 1_000_000_000_000
PRUEFSUMME_MOD = (1 << 63) - 1

_MASKE64 = (1 << 64) - 1
_VORZEICHENBIT64 = 1 << 63


def umbruch64(x):
    """Bildet x auf den Wertebereich eines vorzeichenbehafteten 64-Bit-Ganzworts
    ab: Ueberlauf bricht um, wie in C, Rust (wrapping) und Java."""
    x &= _MASKE64
    return x - (1 << 64) if x & _VORZEICHENBIT64 else x


def pruefsumme(z):
    """summe(z[i] * (i + 1)) mit Ueberlaufumbruch, danach mod 2^63 - 1."""
    summe = 0
    for i in range(GROESSE):
        summe = umbruch64(summe + umbruch64(z[i] * (i + 1)))
    return summe % PRUEFSUMME_MOD


def main():
    # Startbelegung: z[i] = 1_000_000 + i * 37
    z = [1_000_000 + i * 37 for i in range(GROESSE)]

    # Die drei Schritte stehen ausgeschrieben in der Schleife statt in
    # Hilfsfunktionen: 64.000.000 Fortschreibungen mal drei Aufrufe kosten in
    # CPython mehr Zeit als die Rechnung selbst und wuerden die Messung
    # verfaelschen. Ein Python-Programmierer schreibt eine so heisse Schleife
    # ebenso; die Rechenvorschrift ist Zeile fuer Zeile dieselbe.
    #
    # Zwei Eigenschaften der Vorgabe erlauben die kurzen Formen:
    #
    # 1. `klemme(..., 0, 10^12)` haelt jeden Zustandswert dauerhaft in
    #    [0, 10^12], und die Startbelegung liegt schon darin. Alle Operanden
    #    von `%` und `//` sind also nie negativ -- dort stimmen Pythons
    #    abrundende Operatoren mit der gegen null abschneidenden
    #    Ganzzahldivision der Vorgabe ueberein.
    # 2. Damit ist `mal_geteilt(a, b, 10_000)` mit a, b >= 0 gleich
    #    (2*a*b + 10_000) // 20_000 -- das ist genau "auf halbe Betraege von
    #    null weg" gerundet, denn a*b/10_000 = k + f wird zu k + f + 1/2
    #    abgerundet.
    #
    # Kein Zwischenwert verlaesst dabei den 64-Bit-Bereich: das Produkt bleibt
    # unter 10^12 * 10.976, also rund 1,1 * 10^16. Der Umbruch bleibt deshalb
    # der Pruefsumme vorbehalten.
    grenze = GROESSE
    versatz = NACHBAR_VERSATZ
    streuer = STREUER
    nenner = NENNER
    doppel_nenner = 2 * NENNER
    nachbar_teiler = NACHBAR_TEILER
    eigen_teiler = EIGEN_TEILER
    untergrenze = UNTERGRENZE
    obergrenze = OBERGRENZE

    beginn = time.perf_counter_ns()
    for _ in range(SCHRITTE):
        for i in range(grenze):
            # 1. Nachbar -- in-place, ein spaeteres i sieht die schon
            #    fortgeschriebenen Werte der frueheren.
            nachbar = z[(i + versatz) % grenze]
            eigen = z[i]
            # 2. roh = mal_geteilt(z[i], 10_000 + (nachbar mod 977), 10_000)
            zaehler = eigen * (nenner + nachbar % streuer)
            roh = (2 * zaehler + nenner) // doppel_nenner
            # 3. klemme(roh + nachbar/1024 - z[i]/4096, 0, 10^12)
            wert = roh + nachbar // nachbar_teiler - eigen // eigen_teiler
            if wert < untergrenze:
                wert = untergrenze
            elif wert > obergrenze:
                wert = obergrenze
            z[i] = wert
    ende = time.perf_counter_ns()

    print("pruefsumme=%d" % pruefsumme(z))
    print("nanosekunden_je_schritt=%d" % ((ende - beginn) // SCHRITTE))
    print("zustand0=%d" % z[0])


if __name__ == "__main__":
    main()

File: ventures/0016-hedgefonds-simulation-echte-weltwirtschaft/messung-stack/cpp/schritt.cpp
// Deterministischer Festkomma-Weltschritt, C++-Fassung der Stack-Messung.
// Uebersetzt mit: g++ -O2 -std=c++20 schritt.cpp -o schritt
// Kein Gleitkommatyp, keine Fremdbibliothek, keine Eingabe, keine Argumente.

#include <chrono>
#include <cstdint>
#include <cstdio>

namespace {

constexpr int FELDER = 64;
constexpr int SCHRITTE = 1000000;
constexpr std::int64_t OBERGRENZE = 1000000000000LL;
constexpr std::int64_t UNTERGRENZE = 0;

// a * b / c mit 128 Bit im Zwischenergebnis, gerundet auf halbe Betraege von null weg.
inline std::int64_t mal_geteilt(std::int64_t a, std::int64_t b, std::int64_t c) {
    __int128 zaehler = static_cast<__int128>(a) * static_cast<__int128>(b);
    __int128 nenner = static_cast<__int128>(c);
    bool negativ = (zaehler < 0) != (nenner < 0);
    __int128 betrag_zaehler = zaehler < 0 ? -zaehler : zaehler;
    __int128 betrag_nenner = nenner < 0 ? -nenner : nenner;
    __int128 quotient = betrag_zaehler / betrag_nenner;
    __int128 rest = betrag_zaehler % betrag_nenner;
    if (rest * 2 >= betrag_nenner) {
        quotient += 1;
    }
    return static_cast<std::int64_t>(negativ ? -quotient : quotient);
}

inline std::int64_t klemme(std::int64_t x, std::int64_t unten, std::int64_t oben) {
    if (x < unten) return unten;
    if (x > oben) return oben;
    return x;
}

} // namespace

int main() {
    std::int64_t z[FELDER];
    for (int i = 0; i < FELDER; ++i) {
        z[i] = 1000000LL + static_cast<std::int64_t>(i) * 37LL;
    }

    auto beginn = std::chrono::steady_clock::now();
    for (int schritt = 0; schritt < SCHRITTE; ++schritt) {
        for (int i = 0; i < FELDER; ++i) {
            std::int64_t nachbar = z[(i + 17) % FELDER];
            std::int64_t roh = mal_geteilt(z[i], 10000LL + (nachbar % 977LL), 10000LL);
            z[i] = klemme(roh + (nachbar / 1024LL) - (z[i] / 4096LL), UNTERGRENZE, OBERGRENZE);
        }
    }
    auto ende = std::chrono::steady_clock::now();

    // Summe mit Ueberlaufumbruch, danach modulo 2^63 - 1.
    std::uint64_t summe = 0;
    for (int i = 0; i < FELDER; ++i) {
        summe += static_cast<std::uint64_t>(z[i]) * static_cast<std::uint64_t>(i + 1);
    }
    std::uint64_t pruefsumme = summe % ((1ULL << 63) - 1ULL);

    std::int64_t nanosekunden =
        std::chrono::duration_cast<std::chrono::nanoseconds>(ende - beginn).count();
    std::int64_t je_schritt = nanosekunden / SCHRITTE;

    std::printf("pruefsumme=%llu\n", static_cast<unsigned long long>(pruefsumme));
    std::printf("nanosekunden_je_schritt=%lld\n", static_cast<long long>(je_schritt));
    std::printf("zustand0=%lld\n", static_cast<long long>(z[0]));
    return 0;
}
